File: leetcode/56.MergeIntervals.go
package leetcode

import (
	"math"
	"sort"
)

func mergeIntervals(intervals [][]int) [][]int {
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	result := [][]int{
		intervals[0],
	}
	for _, interval := range intervals[1:] {
		start, end := interval[0], interval[1]
		lastEnd := result[len(result)-1][1]
		isOverlapping := start <= lastEnd

		if isOverlapping {
			result[len(result)-1][1] = int(math.Max(float64(lastEnd), float64(end)))
		} else {
			result = append(result, []int{start, end})
		}
	}

	return result
}

File: leetcode/56.MergeIntervals.ts
function merge(intervals: number[][]): number[][] {
  intervals.sort((a, b) => a[0] - b[0])

  let result = [intervals[0]]
  for (const interval of intervals.slice(1)) {
    const [start, end] = interval
    const lastEnd = result[result.length - 1][1]
    const isOverlapping = start <= lastEnd

    if (isOverlapping) {
      result[result.length - 1][1] = Math.max(lastEnd, end)
    } else {
      result.push([start, end])
    }
  }

  return result
}
console.log(
  merge([
    [1, 3],
    [8, 10],
    [15, 18],
    [2, 6],
  ]),
)
